plugins/filter_aws/aws.c

```diff
@@ -182,191 +206,279 @@ static int cb_aws_init(struct flb_filter_instance *f_ins,
+        tag_value_path = flb_sds_create_size(tag_value_path_len);
+        if (!tag_value_path) {
+            flb_errno();
+            return -1;
+        }
+        tmp = flb_sds_printf(&tag_value_path, "%s/%s",
+                             FLB_AWS_IMDS_INSTANCE_TAG,
+                             ctx->tag_keys[i]);
+        if (!tmp) {
+            flb_errno();
+            flb_sds_destroy(tag_value_path);
+            return -1;
+        }
+        tag_value_path = tmp;
+
+        ret = flb_aws_imds_request(ctx->client_imds, tag_value_path, &tag_value,
+                                   &tag_value_len);
+        if (ret < 0) {
+            flb_sds_destroy(tag_value_path);
+            if (ret == -2) {
+                flb_plg_error(ctx->ins, "no value for tag %s", ctx->tag_keys[i]);
+            } else {
+                flb_plg_error(ctx->ins, "could not fetch value for tag %s",
+                              ctx->tag_keys[i]);
+            }
+            return ret;
+        }
 
-    flb_http_client_destroy(client);
-    flb_upstream_conn_release(u_conn);
-    return 0;
-}
+        ctx->tag_values[i] = tag_value;
+        ctx->tag_values_len[i] = tag_value_len;
 
-static int get_metadata(struct flb_filter_aws *ctx, char *metadata_path,
-                        flb_sds_t *metadata, size_t *metadata_len)
-{
-    return get_metadata_by_key(ctx, metadata_path, metadata,
-                               metadata_len, NULL);
+        flb_sds_destroy(tag_value_path);
+    }
+
+    return 0;
 }
 
-/* get VPC metadata, it called IMDS twice.
- * First is for getting the Mac ID and combine into the path for VPC.
- * Second call is using the VPC path to get the VPC id
- */
-static int get_vpc_metadata(struct flb_filter_aws *ctx)
+static int get_ec2_tags(struct flb_filter_aws *ctx)
 {
     int ret;
-    flb_sds_t mac_id = NULL;
-    size_t len = 0;
 
-    /* get EC2 instance Mac id first before getting VPC id */
-    ret = get_metadata(ctx, FLB_FILTER_AWS_IMDS_MAC_PATH, &mac_id, &len);
+    ctx->tags_fetched = false;
 
+    /* get_ec2_tags function might be called multiple times, so we need to always */
+    /* free memory for tags in case of previous allocations */
+    flb_filter_aws_tags_destroy(ctx);
+
+    ret = get_ec2_tag_keys(ctx);
     if (ret < 0) {
-        flb_sds_destroy(mac_id);
-        return -1;
+        return ret;
+    }
+    ret = get_ec2_tag_values(ctx);
+    if (ret < 0) {
+        return ret;
     }
 
-    /* the VPC full path should be like:
-     *latest/meta-data/network/interfaces/macs/{mac_id}/vpc-id/"
-     */
-    flb_sds_t vpc_path = flb_sds_create_size(70);
-    vpc_path = flb_sds_printf(&vpc_path, "%s/%s/%s/",
-                              "/latest/meta-data/network/interfaces/macs",
-                              mac_id, "vpc-id");
-    ret = get_metadata(ctx, vpc_path, &ctx->vpc_id, &ctx->vpc_id_len);
-
-    flb_sds_destroy(mac_id);
-    flb_sds_destroy(vpc_path);
-
-    return ret;
+    ctx->tags_fetched = true;
```

Comment:
In Fluent Bit, we use `int` for boolean and use the values `FLB_TRUE` and `FLB_FALSE`.